python/lsst/ts/envsensors/socket_server.py

```diff
@@ -45,119 +47,85 @@ class SocketServer:
 
     valid_simulation_modes = (0, 1)
 
-    def __init__(self, port, simulation_mode=0):
-        self.port = port
+    def __init__(self, port: int, simulation_mode: int = 0) -> None:
+        self.name = "EnvSensor"
+        self.host: typing.Optional[str] = None  # always read on all interfaces
         if simulation_mode not in self.valid_simulation_modes:
             raise ValueError(
                 f"simulation_mode={simulation_mode} "
                 f"not in valid_simulation_modes={self.valid_simulation_modes}"
             )
 
         self.simulation_mode = simulation_mode
+        self.read_loop_task: asyncio.Future = asyncio.Future()
+        self.log: logging.Logger = logging.getLogger(type(self).__name__)
 
-        self._server = None
-        self._writer = None
-        self._reader = None
-        self._started = False
-        self.log = logging.getLogger(type(self).__name__)
-
-        self.command_handler = CommandHandler(
+        self.command_handler: CommandHandler = CommandHandler(
             callback=self.write, simulation_mode=self.simulation_mode
         )
-
-    async def start(self, keep_running=False):
-        """Start the TCP/IP server.
-
-        Start the command loop and make sure to keep running when instructed to
-        do so.
-
-        Parameters
-        ----------
-        keep_running : bool
-            Used for command line testing and should generally be left to
-            False.
-        """
-        self.log.info("Start called")
-        self._started = True
-        self._server = await asyncio.start_server(
-            self.read_loop, host="127.0.0.1", port=self.port
-        )
-
-        # Request the assigned port from the server so the code starting the
-        # mock controller can use it to connect.
-        if self.port == 0:
-            self.port = self._server.sockets[0].getsockname()[1]
-
-        if keep_running:
-            await self._server.serve_forever()
-
-    async def write(self, data):
+        super().__init__(self.name, self.host, port, self.log, self.connected_callback)
+
+    def connected_callback(self, server: tcpip.OneClientServer) -> None:
+        """A client has connected or disconnected."""
+        if not server.connected:
```

Comment:
I gently suggest testing this only once in this method (you test it twice).

Reply:
Done.